File: lib/games/werewolf/werewolf_server.dart
import 'dart:async';
import 'dart:convert';

import '../../social/host_server.dart';
import 'werewolf_browser.dart';
import 'werewolf_engine.dart';
import 'werewolf_role.dart';

/// Wraps [HostServer] with Werewolf-specific routing. Owns the engine,
/// fans out the right private/public messages, and converts incoming
/// guest commands into engine calls.
class WerewolfServer {
  WerewolfServer({HostServer? server, this.hostName = 'Host'})
      : _server = server ?? HostServer(html: werewolfBrowserHtml);

  final HostServer _server;
  final String hostName;
  final WerewolfEngine engine = WerewolfEngine();

  /// Flutter host plays as this player; not connected over WebSocket.
  static const String hostId = 'host';
  final Map<GuestId, String> _guestToPlayer = {};
  final Map<String, GuestId> _playerToGuest = {};

  final _stateChanges = StreamController<void>.broadcast();
  Stream<void> get onStateChange => _stateChanges.stream;

  StreamSubscription<GuestMessage>? _msgSub;
  StreamSubscription<GuestId>? _leaveSub;

  Future<Uri?> start() async {
    engine.addPlayer(id: hostId, name: hostName, isHost: true);
    final uri = await _server.start();
    _msgSub = _server.onGuestMessage.listen(_onMessage);
    _leaveSub = _server.onGuestLeave.listen(_onLeave);
    _emit();
    return uri;
  }

  Future<void> stop() async {
    await _msgSub?.cancel();
    await _leaveSub?.cancel();
    await _server.stop();
    await _stateChanges.close();
  }

  int get guestCount => _server.guestCount;

  // ---- Host-side actions (called from the Flutter host UI) -------------

  void hostStart() {
    if (!engine.canStart) return;
    engine.start();
    _broadcastPhase();
    _sendRolesPrivately();
    _emit();
  }

  void hostNightAction(String targetId) => _applyNightAction(hostId, targetId);
  void hostDayVote(String? targetId) => _applyDayVote(hostId, targetId);
  void hostHunterShot(String targetId) => _applyHunterShot(hostId, targetId);

  void hostCallTutorialVote() => _openTutorialVote();
  void hostTutorialVote(bool yes) => _submitTutorialVote(hostId, yes);
  void hostDismissTutorial() {
    engine.tutorialVote.markShown();
    _broadcastTutorialState();
    _emit();
  }

  void advanceFromReveal() {
    engine.advanceToDayVote();
    if (engine.phase == WerewolfPhase.gameOver) {
      _broadcastGameOver();
    } else {
      _broadcastPhase();
    }
    _emit();
  }

  // ---- Guest message handling ------------------------------------------

  void _onMessage(GuestMessage msg) {
    final json = msg.asJson;
    final type = json['type'] as String?;
    switch (type) {
      case 'join':
        _handleJoin(msg.from, json);
      case 'night_action':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _applyNightAction(pid, json['targetId'] as String);
      case 'vote':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _applyDayVote(pid, json['targetId'] as String?);
      case 'hunter_shot':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _applyHunterShot(pid, json['targetId'] as String);
      case 'call_tutorial_vote':
        _openTutorialVote();
      case 'tutorial_vote':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _submitTutorialVote(pid, json['yes'] as bool);
    }
  }

  void _onLeave(GuestId guest) {
    final pid = _guestToPlayer.remove(guest);
    if (pid == null) return;
    _playerToGuest.remove(pid);
    engine.removePlayer(pid);
    engine.tutorialVote.removeVoter(pid);
    _broadcastLobby();
    if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) {
      _broadcastTutorialState();
    }
    _emit();
  }

  void _handleJoin(GuestId guest, Map<String, Object?> json) {
    if (engine.phase != WerewolfPhase.lobby) {
      _server.send(guest,
          jsonEncode({'type': 'error', 'message': 'Game already started'}));
      return;
    }
    final name = (json['name'] as String?)?.trim();
    if (name == null || name.isEmpty) return;
    final pid = 'g${_guestToPlayer.length + 1}';
    engine.addPlayer(id: pid, name: name);
    _guestToPlayer[guest] = pid;
    _playerToGuest[pid] = guest;
    _server.send(
        guest,
        jsonEncode({
          'type': 'welcome',
          'yourId': pid,
          'yourName': name,
        }));
    _broadcastLobby();
    _broadcastTutorialState();
    _emit();
  }

  void _applyNightAction(String playerId, String targetId) {
    final p = engine.players[playerId];
    if (p == null || !p.alive) return;
    bool ready = false;
    if (p.role == WerewolfRole.werewolf) {
      ready = engine.submitWolfVote(playerId, targetId);
    } else if (p.role == WerewolfRole.seer) {
      ready = engine.submitSeerTarget(playerId, targetId);
    }
    _emit();
    if (ready) {
      engine.resolveNight();
      _sendSeerResultPrivately();
      _broadcastNightResult();
      if (engine.phase == WerewolfPhase.gameOver) {
        _broadcastGameOver();
      } else if (engine.phase == WerewolfPhase.hunterShot) {
        _broadcastHunterPrompt();
      }
      _emit();
    }
  }

  void _applyDayVote(String playerId, String? targetId) {
    final ready = engine.submitDayVote(playerId, targetId);
    _broadcastVoteUpdate();
    _emit();
    if (ready) {
      engine.resolveDay();
      _broadcastDayResult();
      if (engine.phase == WerewolfPhase.gameOver) {
        _broadcastGameOver();
      } else if (engine.phase == WerewolfPhase.hunterShot) {
        _broadcastHunterPrompt();
      } else {
        _broadcastPhase();
      }
      _emit();
    }
  }

  void _applyHunterShot(String playerId, String targetId) {
    final ok = engine.submitHunterShot(playerId, targetId);
    if (!ok) return;
    _broadcastHunterShotResult();
    if (engine.phase == WerewolfPhase.gameOver) {
      _broadcastGameOver();
    } else if (engine.phase == WerewolfPhase.hunterShot) {
      _broadcastHunterPrompt();
    } else {
      _broadcastPhase();
    }
    _emit();
  }

  // ---- Outbound -------------------------------------------------------

  void _broadcastLobby() {
    _server.broadcast(jsonEncode({
      'type': 'lobby',
      'players': engine.players.values
          .map((p) => {'id': p.id, 'name': p.name, 'isHost': p.isHost})
          .toList(),
      'canStart': engine.canStart,
    }));
  }

  // ---- Tutorial vote ---------------------------------------------------

  void _openTutorialVote() {
    if (engine.phase != WerewolfPhase.lobby) return;
    if (engine.tutorialVote.isOpen) return;
    if (engine.tutorialVote.tutorialShown) return;
    engine.tutorialVote.open(engine.players.keys);
    _broadcastTutorialState();
    _emit();
  }

  void _submitTutorialVote(String voterId, bool yes) {
    if (!engine.tutorialVote.isOpen) return;
    engine.tutorialVote.submit(voterId, yes);
    _broadcastTutorialState();
    _emit();
  }

  void _broadcastTutorialState() {
    final v = engine.tutorialVote;
    _server.broadcast(jsonEncode({
      'type': 'tutorial_vote_state',
      'isOpen': v.isOpen,
      'yesCount': v.yesCount,
      'noCount': v.noCount,
      'eligibleCount': v.eligibleCount,
      'result': v.result,
      'tutorialShown': v.tutorialShown,
    }));
  }

  void _sendRolesPrivately() {
    final wolfIds = engine.players.values
        .where((p) => p.role == WerewolfRole.werewolf)
        .map((p) => p.id)
        .toList();
    for (final p in engine.players.values) {
      final payload = jsonEncode({
        'type': 'role',
        'role': p.role!.name,
        if (p.role == WerewolfRole.werewolf) 'wolfIds': wolfIds,
      });
      final guest = _playerToGuest[p.id];
      if (guest != null) _server.send(guest, payload);
    }
  }

  void _sendSeerResultPrivately() {
    final r = engine.lastSeerResult;
    if (r == null) return;
    final guest = _playerToGuest[r.seerId];
    final payload = jsonEncode({
      'type': 'seer_result',
      'targetId': r.targetId,
      'isWerewolf': r.isWerewolf,
    });
    if (guest != null) _server.send(guest, payload);
  }

  void _broadcastPhase() {
    _server.broadcast(jsonEncode({
      'type': 'phase',
      'phase': engine.phase.name,
      'day': engine.day,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
    }));
  }

  void _broadcastVoteUpdate() {
    _server.broadcast(jsonEncode({
      'type': 'vote_update',
      'votes': engine.dayVotes.map((k, v) => MapEntry(k, v ?? '')),
    }));
  }

  void _broadcastNightResult() {
    final n = engine.lastNight!;
    _server.broadcast(jsonEncode({
      'type': 'phase',
      'phase': engine.phase.name,
      'day': engine.day,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
      'killedId': n.killedId,
    }));
  }

  void _broadcastDayResult() {
    final d = engine.lastDay!;
    _server.broadcast(jsonEncode({
      'type': 'day_result',
      'eliminatedId': d.eliminatedId,
      'tally': d.tally,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
      'eliminatedRole': d.eliminatedId == null
          ? null
          : engine.players[d.eliminatedId]!.role!.name,
    }));
  }

  void _broadcastHunterPrompt() {
    _server.broadcast(jsonEncode({
      'type': 'hunter_prompt',
      'hunterId': engine.pendingHunterShooter,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
    }));
  }

  void _broadcastHunterShotResult() {
    final shots = engine.hunterShotsThisRound;
    if (shots.isEmpty) return;
    final last = shots.last;
    _server.broadcast(jsonEncode({
      'type': 'hunter_shot_result',
      'hunterId': last.hunterId,
      'targetId': last.targetId,
      'targetRole': engine.players[last.targetId]!.role!.name,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
    }));
  }

  void _broadcastGameOver() {
    _server.broadcast(jsonEncode({
      'type': 'game_over',
      'winner': engine.winner!.name,
      'roles': {
        for (final p in engine.players.values) p.id: p.role!.name,
      },
    }));
  }

  Map<String, Object?> _publicPlayer(Player p) =>
      {'id': p.id, 'name': p.name, 'alive': p.alive};

  void _emit() => _stateChanges.add(null);
}

File: android/app/src/main/java/com/example/ubapp/games/werewolf/WerewolfServer.kt
package com.example.ubapp.games.werewolf

import android.content.Context
import com.example.ubapp.social.GuestId
import com.example.ubapp.social.HostServer
import com.example.ubapp.tutorials.GameTutorials
import org.json.JSONArray
import org.json.JSONObject

/** Wraps [HostServer] with Werewolf-specific routing. */
class WerewolfServer(context: Context, val hostName: String = "Host") {
    val engine = WerewolfEngine()
    private val server = HostServer(html = HostServer.htmlAsset(context, "werewolf_browser.html"))
    private val guestToPlayer = HashMap<GuestId, String>()
    private val playerToGuest = HashMap<String, GuestId>()
    var onStateChange: (() -> Unit)? = null

    init { server.onMessage = ::onMessage; server.onLeave = ::onLeave }
    companion object { const val HOST_ID = "host" }

    fun start(): String? {
        engine.addPlayer(HOST_ID, hostName, isHost = true)
        val u = server.startServer(); emit(); return u
    }
    fun stop() = server.stopServer()
    val guestCount: Int get() = server.guestCount

    fun hostStart() { if (!engine.canStart) return; engine.start(); broadcastPhase(); sendRolesPrivately(); emit() }
    fun hostNightAction(targetId: String) = applyNightAction(HOST_ID, targetId)
    fun hostDayVote(targetId: String?) = applyDayVote(HOST_ID, targetId)
    fun hostHunterShot(targetId: String) = applyHunterShot(HOST_ID, targetId)
    fun hostCallTutorialVote() = openTutorialVote()
    fun hostTutorialVote(yes: Boolean) = submitTutorialVote(HOST_ID, yes)
    fun hostDismissTutorial() { engine.tutorialVote.markShown(); broadcastTutorialState(); emit() }
    fun advanceFromReveal() {
        engine.advanceToDayVote()
        if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver() else broadcastPhase()
        emit()
    }

    private fun onMessage(guest: GuestId, raw: String) {
        val j = runCatching { JSONObject(raw) }.getOrNull() ?: return
        when (j.optString("type")) {
            "join" -> handleJoin(guest, j)
            "night_action" -> guestToPlayer[guest]?.let { applyNightAction(it, j.getString("targetId")) }
            "vote" -> guestToPlayer[guest]?.let {
                applyDayVote(it, if (j.isNull("targetId")) null else j.getString("targetId"))
            }
            "hunter_shot" -> guestToPlayer[guest]?.let { applyHunterShot(it, j.getString("targetId")) }
            "call_tutorial_vote" -> openTutorialVote()
            "tutorial_vote" -> guestToPlayer[guest]?.let { submitTutorialVote(it, j.getBoolean("yes")) }
        }
    }

    private fun onLeave(guest: GuestId) {
        val pid = guestToPlayer.remove(guest) ?: return
        playerToGuest.remove(pid)
        engine.removePlayer(pid); engine.tutorialVote.removeVoter(pid)
        broadcastLobby()
        if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) broadcastTutorialState()
        emit()
    }

    private fun handleJoin(guest: GuestId, j: JSONObject) {
        if (engine.phase != WerewolfPhase.LOBBY) {
            send(guest, JSONObject().put("type", "error").put("message", "Game already started")); return
        }
        val name = j.optString("name").trim(); if (name.isEmpty()) return
        val pid = "g${guestToPlayer.size + 1}"
        engine.addPlayer(pid, name)
        guestToPlayer[guest] = pid; playerToGuest[pid] = guest
        send(guest, JSONObject().put("type", "welcome").put("yourId", pid).put("yourName", name).put("game", "werewolf"))
        broadcastLobby(); broadcastTutorialState(); emit()
    }

    private fun applyNightAction(playerId: String, targetId: String) {
        val p = engine.players[playerId] ?: return
        if (!p.alive) return
        val ready = when (p.role) {
            WerewolfRole.WEREWOLF -> engine.submitWolfVote(playerId, targetId)
            WerewolfRole.SEER -> engine.submitSeerTarget(playerId, targetId)
            else -> false
        }
        emit()
        if (ready) {
            engine.resolveNight(); sendSeerResultPrivately(); broadcastNightResult()
            if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
            else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
            emit()
        }
    }

    private fun applyDayVote(playerId: String, targetId: String?) {
        val ready = engine.submitDayVote(playerId, targetId); broadcastVoteUpdate(); emit()
        if (ready) {
            engine.resolveDay(); broadcastDayResult()
            if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
            else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
            else broadcastPhase()
            emit()
        }
    }

    private fun applyHunterShot(playerId: String, targetId: String) {
        if (!engine.submitHunterShot(playerId, targetId)) return
        broadcastHunterShotResult()
        if (engine.phase == WerewolfPhase.GAME_OVER) broadcastGameOver()
        else if (engine.phase == WerewolfPhase.HUNTER_SHOT) broadcastHunterPrompt()
        else broadcastPhase()
        emit()
    }

    private fun broadcastLobby() {
        val arr = JSONArray()
        for (p in engine.players.values) arr.put(JSONObject().put("id", p.id).put("name", p.name).put("isHost", p.isHost))
        broadcast(JSONObject().put("type", "lobby").put("players", arr).put("canStart", engine.canStart))
    }
    private fun sendRolesPrivately() {
        val wolfIds = JSONArray(engine.players.values.filter { it.role == WerewolfRole.WEREWOLF }.map { it.id })
        for (p in engine.players.values) {
            val payload = JSONObject().put("type", "role").put("role", roleName(p.role!!))
            if (p.role == WerewolfRole.WEREWOLF) payload.put("wolfIds", wolfIds)
            playerToGuest[p.id]?.let { send(it, payload) }
        }
    }
    private fun sendSeerResultPrivately() {
        val r = engine.lastSeerResult ?: return
        val guest = playerToGuest[r.seerId] ?: return
        send(guest, JSONObject().put("type", "seer_result").put("targetId", r.targetId).put("isWerewolf", r.isWerewolf))
    }
    private fun broadcastPhase() {
        broadcast(JSONObject()
            .put("type", "phase").put("phase", phaseJson(engine.phase)).put("day", engine.day)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastVoteUpdate() {
        val votes = JSONObject(); for ((k, v) in engine.dayVotes) votes.put(k, v ?: "")
        broadcast(JSONObject().put("type", "vote_update").put("votes", votes))
    }
    private fun broadcastNightResult() {
        val n = engine.lastNight!!
        broadcast(JSONObject()
            .put("type", "phase").put("phase", phaseJson(engine.phase)).put("day", engine.day)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead))
            .put("killedId", n.killedId ?: JSONObject.NULL))
    }
    private fun broadcastDayResult() {
        val d = engine.lastDay!!
        val role = d.eliminatedId?.let { engine.players[it]?.role?.let { r -> roleName(r) } }
        broadcast(JSONObject()
            .put("type", "day_result")
            .put("eliminatedId", d.eliminatedId ?: JSONObject.NULL)
            .put("tally", JSONObject(d.tally as Map<*, *>))
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead))
            .put("eliminatedRole", role ?: JSONObject.NULL))
    }
    private fun broadcastHunterPrompt() {
        broadcast(JSONObject()
            .put("type", "hunter_prompt")
            .put("hunterId", engine.pendingHunterShooter ?: JSONObject.NULL)
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastHunterShotResult() {
        val last = engine.hunterShotsThisRound.lastOrNull() ?: return
        broadcast(JSONObject()
            .put("type", "hunter_shot_result")
            .put("hunterId", last.hunterId).put("targetId", last.targetId)
            .put("targetRole", roleName(engine.players[last.targetId]!!.role!!))
            .put("alive", publicArr(engine.alive)).put("dead", publicArr(engine.dead)))
    }
    private fun broadcastGameOver() {
        val roles = JSONObject()
        for (p in engine.players.values) roles.put(p.id, roleName(p.role!!))
        broadcast(JSONObject().put("type", "game_over")
            .put("winner", if (engine.winner == WerewolfWinner.TOWN) "town" else "werewolves")
            .put("roles", roles))
    }

    private fun openTutorialVote() {
        if (engine.phase != WerewolfPhase.LOBBY) return
        if (engine.tutorialVote.isOpen || engine.tutorialVote.tutorialShown) return
        engine.tutorialVote.open(engine.players.keys); broadcastTutorialState(); emit()
    }
    private fun submitTutorialVote(voterId: String, yes: Boolean) {
        if (!engine.tutorialVote.isOpen) return
        engine.tutorialVote.submit(voterId, yes); broadcastTutorialState(); emit()
    }
    private fun broadcastTutorialState() {
        val v = engine.tutorialVote
        val p = JSONObject().put("type", "tutorial_vote_state")
            .put("isOpen", v.isOpen).put("yesCount", v.yesCount).put("noCount", v.noCount)
            .put("eligibleCount", v.eligibleCount).put("result", v.result ?: JSONObject.NULL)
            .put("tutorialShown", v.tutorialShown)
        if (v.result == true && !v.tutorialShown) {
            p.put("title", GameTutorials.werewolf.title)
            p.put("sections", JSONArray(GameTutorials.werewolf.sectionsJson().map { JSONObject(it as Map<*, *>) }))
            p.put("menuSections", JSONArray(GameTutorials.werewolf.browserMenuSectionsJson().map { JSONObject(it as Map<*, *>) }))
        }
        broadcast(p)
    }

    private fun phaseJson(p: WerewolfPhase) = when (p) {
        WerewolfPhase.LOBBY -> "lobby"; WerewolfPhase.NIGHT -> "night"
        WerewolfPhase.DAY_REVEAL -> "dayReveal"; WerewolfPhase.DAY_VOTE -> "dayVote"
        WerewolfPhase.HUNTER_SHOT -> "hunterShot"; WerewolfPhase.GAME_OVER -> "gameOver"
    }
    private fun roleName(r: WerewolfRole) = r.name.lowercase()

    private fun publicArr(ps: List<WerewolfPlayer>): JSONArray {
        val a = JSONArray()
        for (p in ps) a.put(JSONObject().put("id", p.id).put("name", p.name).put("alive", p.alive))
        return a
    }
    private fun emit() { onStateChange?.invoke() }
    private fun broadcast(o: JSONObject) = server.broadcast(o.toString())
    private fun send(g: GuestId, o: JSONObject) = server.send(g, o.toString())
}
